src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -146,6 +147,88 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 let tp_ty = substs.type_at(0);
                 self.cx().const_usize(self.cx().size_of(tp_ty).bytes())
             }
+            func @ "va_start" | func @ "va_end" => {
+                let va_list = if (tcx.sess.target.target.arch == "aarch64" ||
+                                  tcx.sess.target.target.arch == "x86_64" ||
+                                  tcx.sess.target.target.arch == "powerpc") &&
+                                 !tcx.sess.target.target.options.is_like_windows {
+                    self.load(args[0].immediate(), tcx.data_layout.pointer_align.abi)
+                } else {
+                    args[0].immediate()
+                };
+                let intrinsic = self.cx().get_intrinsic(&format!("llvm.{}", func));
+                self.call(intrinsic, &[va_list], None)
+            }
+            "va_copy" => {
+                if (tcx.sess.target.target.arch == "aarch64" ||
+                                     tcx.sess.target.target.arch == "x86_64" ||
+                                     tcx.sess.target.target.arch == "powerpc") &&
+                                    !tcx.sess.target.target.options.is_like_windows {
+                    let va_list_ty = match tcx.lang_items().va_list() {
+                        Some(did) => {
+                            tcx.type_of(did)
+                        }
+                        None => bug!("Must have the va_list languate item defined")
+                    };
+                    let impl_layout = match va_list_ty.sty {
+                        ty::Adt(def, _) if def.is_struct() => {
+                            let fields = &def.non_enum_variant().fields;
+                            match tcx.type_of(fields[0].did).sty {
+                                ty::Ref(_, element_ty, _) => match element_ty.sty {
+                                    ty::Adt(..) => self.cx().layout_of(element_ty),
+                                    _ => bug!("invalid va_list structure"),
+                                }
+                                _ => bug!("invalid va_list structure")
+                            }
+                        }
+                        _ => bug!("invalid va_list structure"),
+                    };
+                    let impl_ptr = self.alloca(impl_layout.llvm_type(self.cx()), "va_list_impl",
+                                               impl_layout.align.abi);
+                    let intrinsic = self.cx().get_intrinsic(&("llvm.va_copy"));
+                    let src = self.load(args[0].immediate(), tcx.data_layout.pointer_align.abi);
+                    self.call(intrinsic, &[impl_ptr, src], None);
+                    self.load(impl_ptr, impl_layout.align.abi)
+                } else {
+                    let impl_ptr = self.alloca(self.cx().type_i8p(), "va_list_impl",
+                                               tcx.data_layout.pointer_align.abi);
```

Comment:
You don't need to alloca in either case, you have `llresult` to write the destination to (after which you use `return`, I believe.

Reply:
Or `result.llval`, I suppose. You can also access `result.layout` if necessary.

Reply:
Thanks! Much cleaner